src/engraving/libmscore/rendermidi.cpp

```diff
@@ -2473,6 +2473,7 @@ void Score::renderMidi(EventMap* events, bool metronome, bool expandRepeats, con
     ctx.metronome = metronome;
     ctx.renderHarmony = true;
     MidiRenderer(this).renderScore(events, ctx);
+    masterScore()->setExpandRepeats(MScore::playRepeats);
```

Comment:
But we already have a code like this above (line 2470)...

Reply:
No, there we set it as per the `expandRepeats` argument of that method, here as per the preferences setting, the one that is reflected by the Play Repeats button

Reply:
No, there we set it as per the `expandRepeats` argument of that method (from Edit > Preferences > Export > MIDI > Expand repeats) , here as per `MScore::playRepeats`, the setting of the Play Repeats button.

Reply:
The expandRepeats argument should always have the same value as the MScore::playRepeats variable. This method:

`void renderMidi(EventMap* events, bool metronome, bool expandRepeats, const SynthesizerState& synthState);
`

is used in two different places. In the first one:

```
void Score::renderMidi(EventMap* events, const SynthesizerState& synthState)
{
    renderMidi(events, true, MScore::playRepeats, synthState);
}
```

expandRepeats is set explicitly by the MScore::playRepeats value.

In the second one:

```
bool ExportMidi::write(QIODevice* device, bool midiExpandRepeats..)
{
    ...
    m_score->renderMidi(..., midiExpandRepeats, ...);
    ...
}
```

expandRepeats is set by the NotationConfiguration::isPlayRepeatsEnabled() value. The configuration must have the same value as MScore::playRepeats. And this is a bug if they have different values (I would prefer to remove MScore::playRepeats completely but this isn't related to this PR)

Reply:
No, one is a UI setting (the play repeats button in the toolbar), valid for playback, the other a MIDI  export preference setting.
Currentzly theyx interact badly, this PR fixes that and after the MIDI export (which may or may nor have it set) sets it back to the UI setting.

Reply:
If we talk only about the problem and about this method, I think you just need to backup the expandRepeats value and restore after rendering
Will this fix the problem?

Reply:
comparing `expandRepeats` with `MScore::playRepeats` before, storing it (if different) before the `renderScore()` and restoring (if it was different) after should work too, but would be more lines of code for no real benefit.

Reply:
Setting it only if different might be a small optimization?
```
if (expandRepeats != MScore::playRepeats) {
        masterScore()->setExpandRepeats(MScore::playRepeats);
}
```

Reply:
Or maybe resetting it in `ExportMidi::write()`, after the `m_score->renderMidi(..., midiExpandRepeats, ...);`?

Reply:
I'd rather reset it in `renderMidi()` though, close to the set.

Reply:
For example

    bool expandRepeatsBackup = masterScore()->expandRepeats();
    masterScore()->setExpandRepeats(expandRepeats);
    
    MidiRenderer::Context ctx;
    ctx.synthState = synthState;
    ctx.metronome = metronome;
    ctx.renderHarmony = true;
    MidiRenderer(this).renderScore(events, ctx);
    
    masterScore()->setExpandRepeats(expandRepeatsBackup);
    
If a new developer looks at this method, then he will understand what this method does.
I think Roman was confused by this moment. He didn't understand the relationship between expandRepeats and MScore::playRepeats(neither did I)

Reply:
OK, that's just on extra line, and looks OK to me.

Reply:
But... there is no `expandRepeats()`!

Reply:
Back to my `if` clause above...

Reply:
Maybe adding a comment `// like possible if called from ExportMidi::write()`?

Reply:
I think it is better to add a getter for expandRepeats method, we already have the setter

Reply:
OK, done. I liked it better the small and simple way it was before though...

Reply:
The new solution looks much better, thanks!

Reply:
Yes, now it is 3 lines, that's bearable ;-)